libtcmu.c

```diff
@@ -256,6 +256,8 @@ static int set_genl_features(struct nl_sock *sock)
 		goto free_msg;
 
 	ret = nl_send_sync(sock, msg);
+	if (ret < 0)
+		goto free_msg;
```

Comment:
It would be nice to add a log msg about this.

Reply:
ok

Reply:
The goto free_msg is wrong here. nl_send_sync will always free the msg.

So I guess I am saying those whole chunk is not needed.

Reply:
This function frees the msg object after transmitting it by calling nlmsg_free(). It means whether the return is 0 on success or a negative error code, it will calling nlmsg_free to free?

Reply:
https://github.com/thom311/libnl/blob/08160f0555f988db1ec3de65fa188a9d48ebff9d/lib/nl.c#L548

nl_send_sync does:

```
int nl_send_sync(struct nl_sock *sk, struct nl_msg *msg)
{
	int err;

	err = nl_send_auto(sk, msg);
	nlmsg_free(msg);
```

So as long as the function is called nlmsg_free is going to be called. It does not matter if nl_send_sync returns success or failure.

This is why in our code we have that funky goto done which bypasses the nlmsg_free call in our code that is always called after calling nl_send_sync():

```
ret = nl_send_sync(sock, msg);
goto done;
```

Reply:
Thanks, I've seen the source code in libnl.